auth.php

```diff
@@ -1,128 +1,198 @@
+        // token in query string?
+        if ($sig && $sso && $this->validateToken($sig, $sso)) {
+            $this->setTokenCookie(''); // cookie no longer needed
+
+            // get user info from response
+            $ssoResponse = [];
+            parse_str(base64_decode($sso), $ssoResponse);
 
-        if (empty($_SESSION['authdiscourse_login'])) {
-            if (!$this->checkSSO()) {
+            if (!$this->validateResponse($ssoResponse)) {
+                msg($this->getLang('error_login'), -1);
                 return false;
             }
+
+            $userinfo = [
+                'user' => $this->cleanUser($ssoResponse['username']),
+                'name' => $ssoResponse['name'] ?? $ssoResponse['username'],
+                'mail' => $ssoResponse['email'],
+                'grps' => !empty($ssoResponse['groups']) ? explode(',', $ssoResponse['groups']) : [],
+            ];
+            $userinfo['grps'][] = $conf['defaultgroup']; // makes sure users are in default group
+            $userinfo['grps'] = array_map([$this, 'cleanGroup'], $userinfo['grps']);
+
+            // does a local user with this email exist?
+            $found = $this->retrieveUsers(0, 1, ['mail' => '^' . preg_quote_cb($userinfo['mail']) . '$']);
+            if ($found) {
+                // update user with SSO data
+                $oldusername = key($found);
+                if (!$this->modifyUser($oldusername, $userinfo)) {
+                    msg($this->getLang('error_login'), -1);
+                    return false;
+                }
+            } else {
+                // create new user
+                if (!$this->addUser($userinfo)) {
+                    msg($this->getLang('error_login'), -1);
+                    return false;
+                }
+            }
+
+            return $this->storeLogin($userinfo['user'], $userinfo);
         }
 
-        // User is already logged-in or successfully authenticated now.
-        $login = $_SESSION['authdiscourse_login'];
+        // session set? trust it
+        if (isset($_SESSION[DOKU_COOKIE]['auth']['user'])) {
+            return $this->storeLogin($_SESSION[DOKU_COOKIE]['auth']['user'], $_SESSION[DOKU_COOKIE]['auth']['info']);
+        }
 
-        $USERINFO['name'] = $login['name'];
-        $USERINFO['mail'] = $login['email'];
-        $groups = explode(',', $login['groups']);
-        $groups[] = 'user';
-        if ($login['admin'] == 'true') $groups[] = 'admin';
-        if ($login['moderator'] == 'true') $groups[] = 'moderator';
-        $USERINFO['grps'] = $groups;
+        return false;
+    }
 
-        $_SERVER['REMOTE_USER']                = $login['username'];
-        $_SESSION[DOKU_COOKIE]['auth']['user'] = $login['username'];
-        $_SESSION[DOKU_COOKIE]['auth']['info'] = $USERINFO;
+    /**
+     * Saves nonce/token in cookie for comparison with response from SSO provider
+     *
+     * @param string $token
+     */
+    public function setTokenCookie($token)
```

Comment:
My version authenticated the nonce cookie. Are you sure that it's safe to skip authentication?

Reply:
I don't think it matters. When the signature created by discourse matches the calculated signature from the cookie, we can safely assume that the user who sent that cookie to us was also the one who successfully logged into discourse. And that's really all we care about.